public/app/plugins/datasource/prometheus/querybuilder/components/metric-encyclopedia/state/helpers.ts

```diff
@@ -0,0 +1,227 @@
+import { PrometheusDatasource } from 'app/plugins/datasource/prometheus/datasource';
+import { getMetadataHelp, getMetadataType } from 'app/plugins/datasource/prometheus/language_provider';
+
+import { promQueryModeller } from '../../../PromQueryModeller';
+import { regexifyLabelValuesQueryString } from '../../../shared/parsingUtils';
+import { QueryBuilderLabelFilter } from '../../../shared/types';
+import { PromVisualQuery } from '../../../types';
+import { HaystackDictionary, MetricData, MetricsData, PromFilterOption } from '../types';
+
+import { Action, MetricEncyclopediaMetadata, MetricEncyclopediaState } from './state';
+
+export async function getMetadata(
+  datasource: PrometheusDatasource,
+  query: PromVisualQuery
+): Promise<MetricEncyclopediaMetadata> {
+  // Makes sure we loaded the metadata for metrics. Usually this is done in the start() method of the provider but we
+  // don't use it with the visual builder and there is no need to run all the start() setup anyway.
+  if (!datasource.languageProvider.metricsMetadata) {
+    await datasource.languageProvider.loadMetricsMetadata();
```

Comment:
The metadata is always loaded when the `start` method is called on the datasource, so I think it's fair to say that this component doesn't need to init the metadata request, and instead should just dynamically update the hasMetadata based upon the state of the `datasource.languageProvider.metricsMetadata`?

Reply:
Lol I need to read the comments
``` 
Usually this is done in the start() method of the provider but we don't use it with the visual builder and there is no need to run all the start() setup anyway.
```

Ok, now I get why we're doing this, but we're still potentially introducing a race condition. The cache feature would ease the pain here because we can just re-request it without much cost. 

I think this is a bigger problem from ongoing tech debt and not something we should solve here, I've noticed that resource requests don't cancel when we're making duplicate requests, so maybe there's a more robust solution with observables so handle resource requests more like queries.

Reply:
Although in this specific situation we wouldn't want to cancel the request, but wait until that request promise resolves, since metadata doesn't have any request params there's no practical difference between the requests.

Reply:
As long as the metric encyclopedia (now the metrics modal) is inside the metric select, I can remove this. For now the modal option is only shown in the metric select after the initialization of the select options, `getMetrics` which is a function passed into the query builder. 

https://github.com/grafana/grafana/pull/67084/commits/62ee1153dc68ea993e4730e2d8f70dcad916530b

But later, we might want to let the user open the metrics modal before those metrics are loaded... This would be weird though. I don't think that would make sense, it would make more sense to just replace the metric select and `getMetrics` on opening the modal. In the meantime, this works to not have any race conditions, to just let the async metric select load its options and add the metrics modal option at that point.